src/cascade/model/model_writer.py

```diff
@@ -0,0 +1,356 @@
+        smooth_row["n_time"] = age_time_cnt[1]
+        for prior_kind in ["value", "dage", "dtime"]:
+            mulstd_row = prior_table.loc[prior_table.age.isna() & (prior_table.kind == prior_kind)]
+            if all(mulstd_row.assigned):
+                smooth_row[f"mulstd_{prior_kind}_prior_id"] = int(mulstd_row.prior_id)
+        if self._dismod_file.smooth.empty:
+            smooth_row["smooth_id"] = 0
+            self._dismod_file.smooth = self._dismod_file.empty_table("smooth").append(smooth_row, ignore_index=True)
+        else:
+            smooth_row["smooth_id"] = len(self._dismod_file.smooth)
+            self._dismod_file.smooth = self._dismod_file.smooth.append(smooth_row, ignore_index=True)
+        smooth_id = smooth_row["smooth_id"]
+        return smooth_id
+
+    def _add_field_priors(self, grid_name, complete_table):
+        """These are all entries in the priors table for this smooth grid."""
+        # The assigned column will tell us whether mulstds were assigned.
+        complete_table = complete_table.assign(assigned=complete_table.density.notna())
+        complete_table.loc[complete_table.density.isnull(), ["density", "mean", "lower", "upper"]] = \
+            ["uniform", 0, -inf, inf]
+        complete_table = complete_table.assign(density_id=complete_table.density.apply(lambda x: DensityEnum[x].value))
+        # Create new prior IDs that don't overlap.
+        complete_table = complete_table.assign(prior_id=complete_table.index + len(self._dismod_file.prior))
+        complete_table = complete_table.rename(columns={"name": "prior_name"})
+        # Unique, informative names for the priors require care.
+        null_names = complete_table.prior_name.isnull()
+        complete_table.loc[~null_names, "prior_name"] = (
+            complete_table.loc[~null_names, "prior_name"].astype(str) + "    " +
+            complete_table.loc[~null_names, "prior_id"].astype(str)
+        )
+        complete_table.loc[null_names, "prior_name"] = complete_table.loc[
+            null_names, "prior_id"].apply(
+            lambda pid: f"{grid_name}_{pid}"
+        )
+        # Assign age_id and time_id for age and time.
+        complete_table = self._fix_ages_times(complete_table)
+        # Make sure the index still matches the order in the priors list
+        priors_columns = [
+            "prior_id", "prior_name", "lower", "upper", "mean", "std", "eta", "nu", "density_id"
+        ]
+        # Remove columns before saving, but keep extra columns in complete_table for
+        # further construction of grids.
+        prior_table = complete_table.sort_values(by="prior_id").reset_index(drop=True)[priors_columns]
+        if not self._dismod_file.prior.empty:
+            self._dismod_file.prior = self._dismod_file.prior.append(prior_table)
+        else:
+            self._dismod_file.prior = prior_table
+        return complete_table
+
+    def _flush_ages_times_locations(self):
+        if self._flushed:
```

Comment:
So there's no way for the system to become unflushed? Once I've reached this point I can never add data to this writer. Is that guaranteed in any way?

Reply:
The writer defines a protocol that has to be called in order, start to finish. It's a rule. The writer is created, used, and deleted during the session's write_model.